impl/src/main/java/org/jboss/weld/event/ObserverNotifier.java

```diff
@@ -200,9 +201,15 @@ public void fireEvent(Object event, Resolvable resolvable) {
     protected Resolvable buildEventResolvable(Type eventType, Set<Annotation> qualifiers) {
         // We can always cache as this is only ever called by Weld where we avoid non-static inner classes for annotation literals
         Set<Type> typeClosure = sharedObjectCache.getTypeClosureHolder(eventType).get();
-        return new ResolvableBuilder(resolver.getMetaAnnotationStore()).addTypes(typeClosure).addType(Object.class)
+        ResolvableBuilder resolvableBuilder = new ResolvableBuilder(resolver.getMetaAnnotationStore()).addTypes(typeClosure)
+                .addType(Object.class)
                 .addQualifiers(qualifiers)
-                .addQualifierUnchecked(QualifierInstance.ANY).create();
+                .addQualifierUnchecked(QualifierInstance.ANY);
+        if (qualifiers.isEmpty()) {
```

Comment:
I wonder (here and in `EventMetadataImpl`), what if `qualifiers` contains `@Any` (and nothing else)?

These are event qualifiers, so `@Any` is always present (that is, there's no difference between user specifying `@Any` and user _not_ specifying `@Any`), so perhaps `@Default` should be added too in this case?

Reply:
> I wonder (here and in `EventMetadataImpl`), what if `qualifiers` contains `@Any` (and nothing else)?
> 
> These are event qualifiers, so `@Any` is always present (that is, there's no difference between user specifying `@Any` and user _not_ specifying `@Any`), so perhaps `@Default` should be added too in this case?

Why? If you do `@Inject @Any Event<Foo> event` and then `event.fire(new Foo())` then the `@Default` should not be present, or?

Reply:
> I wonder (here and in `EventMetadataImpl`), what if `qualifiers` contains `@Any` (and nothing else)?

Had the same idea and my previous version of this code did exactly what you're suggesting but there were TCKs failures.
IIRC it was the `EventMetadataTest#testSimpleEvent` which doesn't expect `@Default` to be present when you use something like `@Inject @Any Event<SimpleEvent>` and fire event from there.

Reply:
OK, that makes sense. `@Inject Event<Foo>` is equivalent to `@Inject @Default Event<Foo>`, so the set of specified qualifiers is `{@Default}` and the set of event qualifiers is `{@Default, @Any}`.

So the qualifiers here are not _event qualifiers_, they are _specified qualifiers_.

The problem is that

> slightly different behaviour applies to observers with @Default qualifier
> ...
> Such observer will only be notified for events having either no qualifiers or only `@Default` qualifier

talks about _event qualifiers_ (which are never `{}` or `{@Default}`), even though it probably should talk about _specified qualifiers_ (which are also never `{}`, but may be `{@Default}`).

(Technically, that paragraph should probably never have been added. Looking at the history, this was an attempt to clarify that `@Inject Event<Foo>` is equivalent to `@Inject @Default Event<Foo>` due to how injection points work, but that attempt unintentionally added a new provision to the specification that triggers in some weird situation. I actually don't know what that situation is anymore, considering that `BM.getEvent()` should unconditionally add `@Default` to the set of specified qualifiers...)